src/v/datalake/proto_to_arrow_struct.cc

```diff
@@ -0,0 +1,139 @@
+}
+datalake::detail::proto_to_arrow_struct::proto_to_arrow_struct(
+  const google::protobuf::Descriptor* message_descriptor) {
+    using namespace detail;
+    namespace pb = google::protobuf;
+
+    // Set up child arrays
+    _child_arrays.reserve(message_descriptor->field_count());
+    for (int field_idx = 0; field_idx < message_descriptor->field_count();
+         field_idx++) {
+        auto desc = message_descriptor->field(field_idx);
+        switch (desc->cpp_type()) {
+        case pb::FieldDescriptor::CPPTYPE_INT32:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::Int32Type>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_INT64:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::Int64Type>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_BOOL:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::BooleanType>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_FLOAT:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::FloatType>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_DOUBLE:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::DoubleType>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_STRING:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::StringType>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_MESSAGE: {
+            auto field_message_descriptor = desc->message_type();
+            if (field_message_descriptor == nullptr) {
+                throw datalake::initialization_error(
+                  std::string("Can't find schema for nested type : ")
+                  + desc->cpp_type_name());
+            }
+            _child_arrays.push_back(std::make_unique<proto_to_arrow_struct>(
+              field_message_descriptor));
+        } break;
+        default:
+            throw datalake::initialization_error(
+              fmt::format("Unknown type: {}", desc->cpp_type_name()));
+        }
+    }
+    // Make Arrow data types
+
+    // This could be combined into a single loop with the one above, but
+    // this seems more readable to me
+    _fields.reserve(message_descriptor->field_count());
+    for (int field_idx = 0; field_idx < message_descriptor->field_count();
+         field_idx++) {
+        auto field_desc = message_descriptor->field(field_idx);
+        _fields.push_back(_child_arrays[field_idx]->field(field_desc->name()));
+    }
+    _arrow_data_type = arrow::struct_(_fields);
+
+    // Make builder
+    std::vector<std::shared_ptr<arrow::ArrayBuilder>> child_builders;
+    child_builders.reserve(_child_arrays.size());
+    // This could also be collapsed into the above loop
```

Comment:
nit: probably worth removing this comment. In isolation, this reads like a TODO.